### ISSUE TITLE
Implement a "coordinator" binary

### ISSUE THREAD

**Issue by ed255** [appliedzkp/zkevm-circuits#195]
The "coordinator" (this is a provisional name, new ones can be suggested) will be a binary that orchestrates the generation of a proof.
For example, it will take some cli flags or configuration file to determine the geth url to connect to.
Then it will make a series of queries to geth in order to get block, tx and trace data (using the rpc client from bus-mapping).
Then it will feed this data to the CircuitInputBuilder.  Then call the methods on the CircuitInputBuilder to generate the Circuit Input (there may be more than one steps in the future to query more geth data).
Then pass the output of the CircuitInputBuilder to the prover, and finally generate the proof.
In the future, this proof may be sent somewhere.

This is a very basic description of the "coordinator", as the details are still open.

**Comment by spartucus** [comment]
A much more detailed explanation of whole procedure! :)

One nitpicking: The current `bus-mapping` crate and `evm-circuit` crate are separate for now? The test cases in `evm-circuit` do not use `bus-mapping` as witness data input? Am I understand correctly?

**Comment by ed255** [comment]
> One nitpicking: The current `bus-mapping` crate and `evm-circuit` crate are separate for now? The test cases in `evm-circuit` do not use `bus-mapping` as witness data input? Am I understand correctly?

You're correct.  The `bus-mapping` crate and the `zkevm-circuits` crate were developed in parallel and have not been integrated yet.  In the short-term future we plan to integrate and use the structs from the `bus-mapping` as circuit inputs for the `zkevm-circuits`.

**Comment by icemelon** [comment]
I wonder if we should rename `bus-mapping` crate to `circuit-input-builder`. I think the current naming is a bit confusing.

**Comment by lispc** [comment]
I am writing a prover like this <https://github.com/lispc/zkevm-circuits/blob/feature/refactor-step/zkevm-circuits/src/bin/prover.rs>. 

It executes evm bytecodes using the CGO linked geth lib, then builds the circuit inputs for both state circuit and evm circuits, finally verifies the circuit inputs with mock prover.

**Comment by barryWhiteHat** [comment]
@lispc I think this testing infra is a bit different than this issue. 

The goal of this issue is to have a binary to run that will make proofs and submit them to zkrollup on chain. 

I understand yall are working on some testing infra that can be used to test the circuits. 

Am I understanding correctly ? I think that both are important to build btw :) 

Should i make another issue for testing infra ?

**Comment by lispc** [comment]
@barryWhiteHat Yes i understand now. It is indeed more related to test infra.

I added another issue <https://github.com/appliedzkp/zkevm-circuits/issues/204> 

Any comment or direct modification will be helpful for me, thank you!

**Comment by pinkiebell** [comment]
The `coordinator` lives @ `zkevm-chain`. Should I close this issue or is it used for tracking?
